### PR TITLE
Document JIT related OPcache INI settings

### PR DESCRIPTION
**Pull request by cmb69**
Any help on filling out the missing descriptions is appreciated!

### CODE REVIEW

**Comment by talotabi040** [comment]
NRS

File: reference/opcache/ini.xml [hunk 1]
```diff
@@ -845,6 +953,204 @@
...
+    <listitem>
+     <para>
+
+     </para>
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-buffer-size">
+    <term>
+     <parameter>opcache.jit_buffer_size</parameter>
+     <type>integer</type>
```
Comment (nikic) [inline reference/opcache/ini.xml:970]: Drive-by note: We should canonicalize `<type>integer</type>` to `<type>int</type>` and `<type>boolean</type>` to `<type>bool</type>`. We do this correctly for function signatures, but apparently not for many other usages of `<type>`.

Comment (cmb69) [reply reference/opcache/ini.xml:970]: See also https://github.com/php/doc-en/pull/174#discussion_r515692003.

Comment (cmb69) [reply reference/opcache/ini.xml:970]: See https://github.com/php/doc-en/pull/176 and https://github.com/php/doc-base/pull/2.

File: reference/opcache/ini.xml [hunk 2]
```diff
@@ -845,6 +953,204 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-prof-threshold">
+    <term>
+     <parameter>opcache.jit_prof_threshold</parameter>
+     <type>float</type>
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1007]: When using the "profile on first request" trigger mode, this threshold determines whether a function is considered hot. The number of calls to the function divided by the number of calls to all functions must be above the threshold. For example, a threshold of `0.005` means that functions that made up more than `0.5%` of all calls will the JIT compiled.

File: reference/opcache/ini.xml [hunk 3]
```diff
@@ -845,6 +953,204 @@
...
+     </para>
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-loops-unroll">
+    <term>
+     <parameter>opcache.jit_max_loops_unroll</parameter>
+     <type>OnUpdateUnrollL</type>
```
Comment (nikic) [inline reference/opcache/ini.xml:1113]:
```suggestion
     <type>int</type>
```

File: reference/opcache/ini.xml [hunk 4]
```diff
@@ -845,6 +953,204 @@
...
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1117]: Maximum number of loops that can be unrolled within one trace.

(TBH I'm not totally sure what that means, is this about distinct loops or loop iterations?)

File: reference/opcache/ini.xml [hunk 5]
```diff
@@ -845,6 +953,204 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-recursive-calls">
+    <term>
+     <parameter>opcache.jit_max_recursive_calls</parameter>
+     <type>OnUpdateUnrollL</type>
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1128]: Maximum number of recursive calls that can be included within one trace.

File: reference/opcache/ini.xml [hunk 6]
```diff
@@ -845,6 +953,204 @@
...
+     </para>
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-recursive-calls">
+    <term>
+     <parameter>opcache.jit_max_recursive_calls</parameter>
+     <type>OnUpdateUnrollL</type>
```
Comment (nikic) [inline reference/opcache/ini.xml:1124]:
```suggestion
     <type>int</type>
```

File: reference/opcache/ini.xml [hunk 7]
```diff
@@ -845,6 +953,204 @@
...
+    <listitem>
+     <para>
+
+     </para>
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-recursive-return">
+    <term>
+     <parameter>opcache.jit_max_recursive_returns</parameter>
+     <type>OnUpdateUnrollL</type>
```
Comment (nikic) [inline reference/opcache/ini.xml:1135]:
```suggestion
     <type>int</type>
```

File: reference/opcache/ini.xml [hunk 8]
```diff
@@ -845,6 +953,204 @@
...
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1139]: Maximum number of recursive returns that may be included within one trace.

File: reference/opcache/ini.xml [hunk 9]
```diff
@@ -845,6 +953,204 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-polymorphic-calls">
+    <term>
+     <parameter>opcache.jit_max_polymorphic_calls</parameter>
+     <type>integer</type>
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1150]: Maximum number of polymorphic (dynamic) calls to include within one trace.

File: reference/opcache/ini.xml [hunk 10]
```diff
@@ -845,6 +953,204 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-blacklist-root-trace">
+    <term>
+     <parameter>opcache.jit_blacklist_root_trace</parameter>
+     <type>integer</type>
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1095]: Maximum number of times the compilation of a root trace is attempted before it is blacklisted.

File: reference/opcache/ini.xml [hunk 11]
```diff
@@ -845,6 +953,204 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-blacklist-side-trace">
+    <term>
+     <parameter>opcache.jit_blacklist_side_trace</parameter>
+     <type>integer</type>
+    </term>
+    <listitem>
+     <para>
+
```
Comment (nikic) [inline reference/opcache/ini.xml:1106]: Maximum number of times the compilation of a side trace is attempted before it is blacklisted.

File: reference/opcache/ini.xml (hunk 4 above)
Comment (cmb69) [reply reference/opcache/ini.xml:1117]: Maybe @dstogov can clarify?

**Comment by cmb69** [comment]
Thanks a lot, @nikic!

File: reference/opcache/ini.xml [hunk 12]
```diff
@@ -845,6 +953,278 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-polymorphic-calls">
+    <term>
+     <parameter>opcache.jit_max_polymorphic_calls</parameter>
+     <type>integer</type>
+    </term>
+    <listitem>
+     <para>
+      Maximum number of polymorphic (dynamic) calls to include within one trace.
```
Comment (dstogov) [inline reference/opcache/ini.xml:1224]: Maximum number of attempts to inline polymorphic (dynamic or method) call. Calls above this limit treated as megamorphic and not inlined.

File: reference/opcache/ini.xml [hunk 13]
```diff
@@ -845,6 +953,278 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-recursive-return">
+    <term>
+     <parameter>opcache.jit_max_recursive_returns</parameter>
+     <type>int</type>
+    </term>
+    <listitem>
+     <para>
+      Maximum number of recursive returns that may be included within one trace.
```
Comment (dstogov) [inline reference/opcache/ini.xml:1213]: Maximum number of unrolled recursive return loops.

File: reference/opcache/ini.xml [hunk 14]
```diff
@@ -845,6 +953,278 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-recursive-calls">
+    <term>
+     <parameter>opcache.jit_max_recursive_calls</parameter>
+     <type>int</type>
+    </term>
+    <listitem>
+     <para>
+      Maximum number of recursive calls that can be included within one trace.
```
Comment (dstogov) [inline reference/opcache/ini.xml:1202]: Maximum number of unrolled recursive call loops.

File: reference/opcache/ini.xml [hunk 15]
```diff
@@ -845,6 +953,278 @@
...
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-loop-unrolls">
+    <term>
+     <parameter>opcache.jit_max_loops_unroll</parameter>
+     <type>int</type>
+    </term>
+    <listitem>
+     <para>
+      Maximum number of loops that can be unrolled within one trace.
```
Comment (dstogov) [inline reference/opcache/ini.xml:1191]: Maximum number of attempts to unroll a loop in a side trace, trying to reach the root trace and close the outer loop.

File: reference/opcache/ini.xml (hunk 12 above)
Comment (nikic) [reply reference/opcache/ini.xml:1224]: Just to check, this limits the number of side-traces from a single polymorphic call, right?

Comment (dstogov) [reply reference/opcache/ini.xml:1224]: Human language is not the best one :)

For example, with "jit_max_polymorphic_calls=2", "$obj->foo()" is going to be inlined for the first occurred class of $obj, then if we get object of another class, a side trace with another inlined method is going to be compiled, then if we get object of third class, another side trace starts, but method is not going to be inlined anymore. Trace is terminated by the transfer to the called method (goto *opline->handler).

**Comment by cmb69** [comment]
Thanks, @dstogov! I have incorporated your suggestions. 

I think this good to be merged now.

File: reference/opcache/ini.xml [hunk 16]
```diff
@@ -815,6 +923,280 @@
...
+     <para>
+      A bit mask specifying which JIT debug output to enable.
+      For possible values, please consult <filename>zend_jit.h</filename>.
+     </para>
+    </listitem>
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-bisect-limit">
+    <term>
+     <parameter>opcache.jit_bisect_limit</parameter>
+     <type>integer</type>
```
[the remark this answers is not in the document]
Comment (nikic) [reply reference/opcache/ini.xml:1031]: Here and in a few more places.

File: reference/opcache/ini.xml [hunk 17]
```diff
@@ -815,6 +923,280 @@
...
+   </varlistentry>
+   <varlistentry xml:id="ini.opcache.jit-max-polymorphic-calls">
+    <term>
+     <parameter>opcache.jit_max_polymorphic_calls</parameter>
+     <type>integer</type>
+    </term>
+    <listitem>
+     <para>
+      Maximum number of attempts to inline polymorphic (dynamic or method) calls.
+      Calls above this limit treated as megamorphic and are not inlined.
```
Comment (nikic) [inline reference/opcache/ini.xml:1196]:
```suggestion
      Calls above this limit are treated as megamorphic and are not inlined.
```

**Review by nikic — APPROVED** [review]

**Comment by cmb69** [comment]
Thanks Nikita!

Applied as http://svn.php.net/viewvc?view=revision&revision=351205.
